rfcs/0000-rlp-v2.md

```diff
@@ -130,90 +135,87 @@ spec:
+
+In this example, a distinct limit will be associated ("bound") to each individual HTTPRouteRule of the targeted HTTPRoute, by using the `triggers` field for selecting (or "sub-targeting") the HTTPRouteRule.
 
-In this example, a distinct limit will be associated to each individual rule of the targeted HTTPRoute, using the `rules` field for fine-grained filtering (or "sub-targeting").
-- `GET|POST /toys*` → 50rpm, enforced per username (counter qualifier) and only in case the user is not an admin (condition).
+The following limit definitions will be bound to each HTTPRouteRule:
+- `GET|POST /toys*` → 50rpm, enforced per username (counter qualifier) and only in case the user is not an admin ("soft" condition).
 - `/assets/*` → 5rpm / 100rp12h
 
-Each rule in the RLP must perfectly match a rule in the HTTPRoute. In case there is no identical match amongst the HTTPRoute rules for a given rule defined in the RLP, the RLP is considered invalid. In case there is more than one rule specified in the RLP that is an identical match to the same rule in the HTTPRoute, the first matching rule on the list in the RLP is bound to its identical match in the HTTPRoute, thus "shadowing" any other rule on the list that is also an identical match.
+Each set of trigger matches in the RLP will be matched to **all** HTTPRouteRules whose HTTPRouteMatches is a superset of the set of trigger matches in the RLP. For every HTTPRouteRule matched, the HTTPRouteRule will be bound to the corresponding limit definition that specifies that trigger. In case no HTTPRouteRule is found containing at least one HTTPRouteMatch that is identical to some set of matching rules of a particular limit definition, the limit definition is considered invalid and reported as such in the status of RLP.
 
 ```yaml
 apiVersion: kuadrant.io/v2beta1
 kind: RateLimitPolicy
 metadata:
   name: toystore-per-endpoint
+  namespace: toystore
 spec:
   targetRef:
     group: gateway.networking.k8s.io
     kind: HTTPRoute
     name: toystore
   limits:
-    - name: toys
-      rules:
-        - matches: # matches the 1st rule in the HTTPRoute
-            - path:
-                type: PathPrefix
-                value: "/toys"
-              method: GET
-            - path:
-                type: PathPrefix
-                value: "/toys"
-              method: POST
+    toys:
       rates:
-        - limit: 50
-          duration: 1
-          unit: minute
+      - limit: 50
+        duration: 1
+        unit: minute
       counters:
-        - auth.identity.username
+      - auth.identity.username
+      triggers:
+      - matches: # matches the 1st HTTPRouteRule (i.e. GET or POST to /toys*)
+        - path:
+            type: PathPrefix
+            value: "/toys"
+          method: GET
+        - path:
+            type: PathPrefix
+            value: "/toys"
+          method: POST
```

Comment:
Would: 

```yaml
      - matches: 
        - path:
            type: PathPrefix
            value: "/toys"
```

also `matches the 1st HTTPRouteRule (i.e. GET or POST to /toys*)` ? It would right, as this match is a subset of the two matchers in the route... am I understanding this right?

Reply:
I was thinking on atomic HTTPRouteMatches at first, so no. But maybe yes?

Reply:
Oh ok, I misread that then... You mean any _complete_ match will target the rule, so that if I have match A & match B to Rule 1, I don't need to target A & B, but A or B suffices... but either needs to be _complete_, right?

Reply:
I mean _any subset of the HTTPRouteMatches_, not _any subset of **a** HTTPRouteMatch_.

In the example,

```yaml
- matches:
  - path:
      type: PathPrefix
      value: "/toys"
    method: GET
  - path:
      type: PathPrefix
      value: "/toys"
    method: POST
``` 

are 2 HTTPRouteMatches.

The set of all subsets contains 2<sup>2</sup> = 2 × 2 = 4 elements:

- `[{}]`
- `[{ path: { type: PathPrefix, value: /toys }, method: GET }]`
- `[{ path: { type: PathPrefix, value: /toys }, method: POST }]`
- `[{ path: { type: PathPrefix, value: /toys }, method: GET }, { path: { type: PathPrefix, value: /toys }, method: POST }]`

Reply:
> if I have match A & match B to Rule 1, I don't need to target A & B, but A or B suffices... but either needs to be _complete_, right?

Correct.

If you have:

```yaml
- matches:
  - path:
      type: PathPrefix
      value: "/toys"
    method: GET
  - path:
      type: PathPrefix
      value: "/toys"
    method: POST
```

Then just

```yaml
- matches:
  - path:
      type: PathPrefix
      value: "/toys"
    method: GET
```

or just 

```yaml
- matches:
  - path:
      type: PathPrefix
      value: "/toys"
    method: POST
```

would suffice.

Reply:
I re-read this and now need to step back and change what I said before.

To be possible for policy owners to define triggers in the policy that are strictly about what's intended to be protected, we have to allow for a subset of an HTTPRouteMatch. The contrary (i.e. forcing only atomic HTTPRouteMatches in the triggers) would cause the user sometimes having to add to the triggers more than intended, which is exactly what we wanted to avoid when we decided to move away from defining triggers in terms of full HTTPRouteRules.

With that, based on the example from before, one who writes the following trigger matches:

```yaml
triggers:
- matches:
  - path:
      type: PathPrefix
      value: "/toys"
```

would have written a **valid** trigger and binding the limit definition to the HTTPRouteRule:

```yaml
rules:
- matches:
  - path:
      type: PathPrefix
      value: "/toys"
    method: GET
  - path:
      type: PathPrefix
      value: "/toys"
    method: POST
```

What's put in the triggers nonetheless still have to be explicitly stated in the HTTPRouteRules. E.g. the following trigger:

```yaml
triggers:
- matches:
  - path:
      type: Exact
      value: "/toys/special"
```

would be an **invalid** trigger even if it's technically a subset of the HTTPRouteRule above.

<br/>

This is the opposite of what I had said. I will make sure the text is clear about that.

Thanks @alexsnaps!

Reply:
Done. _Example 2_ modified to make it somewhat clearer that a subset of an HTTPRouteMatch (as long as explicitly stated in the HTTPRoute) is also acceptable.